Sources/SourceKit/clangd/ClangLanguageServer.swift

```diff
@@ -24,22 +24,125 @@ final class ClangLanguageServerShim: ToolchainLanguageServer {
 
-  let buildSystem: BuildSystem
+  private let buildSystem: BuildSystem
 
-  let clang: AbsolutePath?
+  private let clangdPath: AbsolutePath
 
-  /// Creates a language server for the given client using the sourcekitd dylib at the specified path.
-  public init(client: Connection, clangd: Connection, buildSystem: BuildSystem,
-              clang: AbsolutePath?) throws {
-    self.clangd = clangd
-    self.buildSystem = buildSystem
-    self.clang = clang
+  private let clangdOptions: [String]
+
+  private let client: MessageHandler
+
+  private var state: LanguageServerState {
+     didSet {
+       for handler in stateChangeHandlers {
+         handler(oldValue, state)
+       }
+     }
+   }
+
+  private var stateChangeHandlers: [(_ oldState: LanguageServerState, _ newState: LanguageServerState) -> Void] = []
+
+  /// Ask the parent of this language server to re-open all documents in this language server.
+  private let reopenDocuments: (ToolchainLanguageServer) -> Void
+
+  /// The `InitializeRequest` with which `clangd` was originally initialized.
+  /// Stored so we can replay the initialization when clangd crashes.
+  private var initializeRequest: InitializeRequest?
+
+  public init(client: MessageHandler,
+              clangdPath: AbsolutePath,
+              buildSettings: BuildSystem?,
+              clangdOptions: [String],
+              reopenDocuments: @escaping (ToolchainLanguageServer) -> Void
+  ) throws {
+    self.client = client
+    self.buildSystem = buildSettings ?? BuildSystemList()
+    self.clangdPath = clangdPath
+    self.clangdOptions = clangdOptions
+    self.state = .connected
+    self.reopenDocuments = reopenDocuments
+  }
+
+  func initialize() throws {
+    let clientToServer: Pipe = Pipe()
+    let serverToClient: Pipe = Pipe()
+
+    let connection = JSONRPCConnection(
+      protocol: MessageRegistry.lspProtocol,
+      inFD: serverToClient.fileHandleForReading.fileDescriptor,
+      outFD: clientToServer.fileHandleForWriting.fileDescriptor
+    )
+
+    self.clangd = connection
+
+    connection.start(receiveHandler: client)
+
+    let process = Foundation.Process()
+
+    if #available(OSX 10.13, *) {
+      process.executableURL = clangdPath.asURL
+    } else {
+      process.launchPath = clangdPath.pathString
+    }
+
+    process.arguments = [
+      "-compile_args_from=lsp",   // Provide compiler args programmatically.
+      "-background-index=false",  // Disable clangd indexing, we use the build
+      "-index=false",             // system index store instead.
+    ] + clangdOptions
+
+    process.standardOutput = serverToClient
+    process.standardInput = clientToServer
+    process.terminationHandler = { [weak self] process in
+      log("clangd exited: \(process.terminationReason) \(process.terminationStatus)")
+      connection.close()
+      if process.terminationStatus != 0 {
+        if let self = self {
+          self.state = .connectionInterrupted
+          self.restartClangd()
```

Comment:
I'm concerned that this could crash loop. For sourcekitd, XPC prevents us from re-spawning too fast in a loop. We should implement something here to delay the relaunch if we crash loop (e.g. after the second crash in a short time start delaying a few seconds before trying again).

Reply:
FWIW VSCode will restart an LSP for up to 5 times within a 3 minute interval. Really depends upon the type of crash - most of which I've seen are persistent or related to a single document.

Might make sense to do something similar here

Reply:
Good point. I have added some super simple crash-loop prevention code: If `clangd` has crashed in the last 30 seconds, delay any further restart by 10 seconds. I think this should be sufficient to prevent us from going wild im crash loop scenarios. I also added a test case that checks that `clangd` isn’t restarted twice in a row within 5 seconds.